xarray/core/utils.py

```diff
@@ -1197,6 +1198,49 @@ def _resolve_doubly_passed_kwarg(
     return kwargs_dict
 
 
+def attempt_import(module: str) -> ModuleType:
+    """Import an optional dependency, and raise an informative error on failure.
+
+    Parameters
+    ----------
+    module : str
+        Module to import. For example, ``'zarr'`` or ``'matplotlib.pyplot'``.
+
+    Returns
+    -------
+    module : ModuleType
+        The Imported module.
+
+    Raises
+    ------
+    ImportError
+        If the module could not be imported.
+
+    Notes
+    -----
+    Static type checkers will not be able to infer the type of the returned module,
+    so it is recommended to precede this function with a direct import of the module,
+    guarded by an ``if TYPE_CHECKING`` block, to preserve type checker functionality.
+    """
+    install_mapping = dict(nc_time_axis="nc-time-axis")
+    package_purpose = dict(
+        zarr="for working with Zarr stores",
+        cftime="for working with non-standard calendars",
+        matplotlib="for plotting",
+        hypothesis="for the `xarray.testing.strategies` submodule",
+    )
+    package_name = module.split(".")[0]  # if submod, get the top-level package name
+
+    if module_available(package_name):
+        return importlib.import_module(module)
+    install_name = install_mapping.get(package_name, package_name)
+    reason = package_purpose.get(package_name, "")
+    raise ImportError(
```

Comment:
Not sure if we should try to import the package and then add the traceback to this error.
Might have some additional useful information.

Reply:
Sure, if we do that then I think we should just use a try-except block instead of integrating the `attempt_import` function? e.g. something like:

```python
    try:
        return importlib.import_module(module)
    except ImportError as e:
        raise ImportError(
            f"The {install_name} package is required {reason} but could not be imported."
            " Please install it with your package manager (e.g. conda or pip).\n"
            f"Traceback: {e}"
        ) from e
```

```python
attempt_import("foo")
```

```python
*** ImportError: The foo package is required  but could not be imported. Please install it with your package manager (e.g. conda or pip).
Traceback: No module named 'foo'
```

What do you think?

Reply:
Would just having `raise ... from e` suffice, rather than adding the traceback into the string? Then python should handle the concat of the errors, iiuc?

Reply:
Ahh, yes you are right 😅 I just pushed [9a41073](https://github.com/pydata/xarray/pull/9561/commits/9a41073dacb88bacfbad046a4affa64c0c2fecad), LMK if you think this is preferable to the previous design!